ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SubjectComparisonVerifier.kt

```diff
@@ -0,0 +1,265 @@
+                throw SAMLComplianceException.create(SAMLCore_3_3_4_b2,
+                        message = "One of the Response's Subject identifier's Format attribute " +
+                                "[${resFormat ?: UNSPECIFIED_URI}] is not identical to the " +
+                                "AuthnRequest's Subject identifier's Format attribute " +
+                                "[$reqFormat].",
+                        node = resId)
+        }
+
+        val reqAttributes = reqId.attributeList().filter { it.localName != FORMAT }.toSet()
+        val resAttributes = resId.attributeList().filter { it.localName != FORMAT }.toSet()
+
+        if (reqAttributes != resAttributes)
+            throw SAMLComplianceException.create(SAMLCore_3_3_4_b2,
+                    SAMLCore_3_4_1_4_b1,
+                    message = "One of the Response'a Subject identifier's attributes " +
+                            "[$resAttributes] are not identical to the" +
+                            "AuthnRequest's Subject identifier's attributes [$reqAttributes].",
+                    node = resId)
+    }
+
+    /**
+     * Compares the text content of the identifiers. Special checking due to nameIdPolicyFormat.
+     */
+    private fun verifyIdContentsMatchAuthnRequest(reqId: Node, resId: Node) {
+        val reqFormat = reqId.filteredFormatValue()
+        val resFormat = resId.filteredFormatValue()
+
+        // If they share a format value that isn't "unspecified", but have different contents
+        if (reqFormat == resFormat
+                && reqFormat != null
+                && reqId.textContent != resId.textContent) {
+            throw SAMLComplianceException.create(SAMLCore_3_3_4_b2,
+                    SAMLCore_3_4_1_4_b,
+                    message = "One of the Response's Subject identifier's content " +
+                            "[${resId.textContent}] is not identical to the AuthnRequest's " +
+                            "Subject identifier's content [${reqId.textContent}], even though " +
+                            "they share the same identifier format [$reqFormat].",
+                    node = resId)
+        }
+    }
+
+    /**
+     * Verifies that the request SubjectConfirmations **strongly match** the response confirmations.
+     *
+     * Note that this matching only compares the methods of the SubjectConfirmations. In order to
+     * properly strongly match we would need to "confirm" the SubjectConfirmation (Hard To Test)
+     */
+    private fun verifyConfirmationsMatchAuthnRequest(requestConfirmations: List<Node>,
+                                                     responseConfirmations: List<Node>) {
+
+        val requestMethods = requestConfirmations
+                .map { it.attributeText("Method") }
+                .filter { it != null }.toSet()
+        val responseMethods = responseConfirmations
+                .map { it.attributeText("Method") }
+                .filter { it != null }.toSet().toSet()
```

Comment:
✏️ For this line and the usage above, you can just call `filterNotNull()` instead of `filter { it != null }`.